src/NuGet.Core/NuGet.Protocol/Resources/ServiceIndexResourceV3.cs

```diff
@@ -105,7 +111,25 @@ private IReadOnlyList<ServiceIndexEntry> GetBestVersionMatchForType(NuGetVersion
             else
             {
                 // Find all entries with the same version.
-                return entries.Where(e => e.ClientVersion == bestMatch.ClientVersion).ToList();
+                List<ServiceIndexEntry> serviceIndexEntries = new List<ServiceIndexEntry>();
+
+                foreach (var entry in entries)
```

Comment:
Shouldn't we be doing this in MakeLookup, ensuring it's done once rather than everytime?